### DISCUSSION TITLE
Generating static page HTML based on JSON (with middleware)

### DISCUSSION THREAD

**Discussion by andreassoegaard** [nuxt/nuxt#9717] [Questions (Nuxt 2)]
**TLDR**

- We generate all our routes of our SaaS solution in router.extendRoutes() in nuxt.config, with JSON data put in the meta object of the routes, used to render the layout of the page component.
- I'm seeking help/advice to statically generate the routes - or at least speed them up - to avoid/minimize rendering time.
- All routes are protected by a middleware based on `nuxt-auth` module (could be a problem?).

**LONG READ**

I made a page builder in Nuxt, which we use internally to build the routes of our SaaS and the contents inside of them (input fields, selects, radio buttons, etc.).

This page builder generates a lot of JSON for each page. I fetch the page data and JSON from our API and put it into the meta object of the routes in router.extendRoutes() in nuxt.config, to access it in my page component like this:

**page.vue** (the page component of all routes)
```
<b-tabs>
  <b-tab v-for="(tab, index) in $route.meta.json" :key="index" :title="tab.tab !== '' ? tab.tab : 'Content'">
    <page-builder :tab-index="index" />
  </b-tab>
</b-tabs>
```

Inside the `<page-builder :tab-index="index" />` component i have more loops to take care of the rows, columns, boxes, fields. So a lot of loops inside loops to create the layout. You can see above that I start out by looping tabs. And inside each tab, `<page-builder />` takes care of looping rows, columns, boxes, contents, etc.

So a lot of frontend power is currently used to render the page layouts, which I'm obviously trying to minimize/avoid.

I turned on static mode in `nuxt.config` in order to try and speed up the pages using `npm run generate`. But I see in the `/dist` folder that no HTML are actually generated for my routes. Only some .js files are loaded into the body:

```
<body>
  <div id="__nuxt"></div>
  <script defer src="/_nuxt/static/1630067976/cockpit/state.js"></script>
  <script src="/_nuxt/eed817d.js" defer></script>
  <script src="/_nuxt/fa770a0.js" defer></script>
  <script src="/_nuxt/d0f9975.js" defer></script>
  <script src="/_nuxt/677bce9.js" defer></script>
</body>
```

All routes are by the way protected by a middleware, that checks if you are logged in, using `nuxt-auth` package. I wonder if that affects it? Because I noticed that the HTML of the log-in page (obviously not protected by a middleware) is actually generated...

- Do you have any tips, tricks and/or ideas to how I could optimize this, to make my routes as static/fast as possible?
- Maybe you can suggest another, better approach?
- Maybe it is not possible to do a static generation of pages in a protected environment with middlewares?
- Maybe I should use `middlewares` and `nuxt-auth` differently?

Any comments are appreciated - tell me if you have any questions!

Have a great day! 😄
